### ISSUE TITLE
Get a Serial No back on stock (without Sales Return)

### ISSUE THREAD

**Issue by stacq85** [frappe/erpnext#21523]
For repair purposes, we would like to get Items already delivered to Customers (using Delivery Notes) located on an internal ERPNext warehouse.
We don’t want to use the Sales Return option, because it affects the Sales Order status (status becomes: 'Partly Delivered'). After a Sales Order is completed, we don’t want to change it anymore.

Using a Stock Entry (Material Receipt) doesn’t work. Then the following error message pops up: `Serial No xxxxxx has already been received.`

Because the Item has to be returned to the Customer, a Delivery Note has to be created when the Item is repaired. As far as I know, I can’t create a Delivery Note without having the Item on an internal Stock location. 

We would like to have an option to get a Serial No (Item) back on an internal stock location without affecting the Sales Order or Delivery Note linked to the Serial No.

**Comment by douglas-watson** [comment]
I would like to add a vote for this feature request.

We also manage repairs and refurbishing of serialized items. It's important for us to return the items to stock, as this allows us to track refurbishing using the Manufacturing module.

Sales Returns don't match our workflow, for the same reason as stiemi.

**Workaround**:

We've actually just noticed this problem, after two years of using ERPNext, because until now we delivered these items to customer warehouses. We can move the items back from the customer warehouse to our own warehouses. This approach leads to other issues though (namely accounting), so we want to move away from it.

**Comment by douglas-watson** [comment]
I had a look at the code that generates the error, and I have the impression that this behaviour may be a bug.

From stock/doctype/serial_no/serial_no.py (https://github.com/frappe/erpnext/blob/a6f98d48bcb59a7e89030d596dae3b102b887981/erpnext/stock/doctype/serial_no/serial_no.py)
```python
def validate_serial_no(sle, item_det):
...
					if cint(sle.actual_qty) > 0 and has_duplicate_serial_no(sr, sle):
						frappe.throw(_("Serial No {0} has already been received").format(serial_no),
							SerialNoDuplicateError)
```

The intention of this code seems to be to avoid duplicate serial numbers, i.e. to avoid a new item entering the stock with the same serial no as an existing item in the stock. In the repair scenario, the item returning to stock is _not a duplicate_, but the very same item. This should be allowed, as it doesn't induce the problems that result from duplicates.

The duplicate detection logic currently reads:
```python
def has_duplicate_serial_no(sn, sle):
	if (sn.warehouse and not sle.skip_serial_no_validaiton
		and sle.voucher_type != 'Stock Reconciliation'):
		return True

	if sn.company != sle.company:
		return False

	status = False
	if sn.purchase_document_no:
		if sle.voucher_type in ['Purchase Receipt', 'Stock Entry', "Purchase Invoice"] and \
			sn.delivery_document_type not in ['Purchase Receipt', 'Stock Entry', "Purchase Invoice"]:
			status = True

		if status and sle.voucher_type == 'Stock Entry' and \
			frappe.db.get_value('Stock Entry', sle.voucher_no, 'purpose') != 'Material Receipt':
			status = False

	return status
```

It could be modified to allow a Materials Receipt for a serial number that currently has been delivered by a Delivery Note, i.e. if "delivery_document_type" is "Delivery Note", the item should not be a duplicate. This should work even if the delivery note has transferred the item to a customer warehouse, as in this specific case the delivery note is listed as the item's purchase_document.


To summarize:

# How to reproduce the issue:

- Issue a delivery note for a serialised item
- Create a Stock Entry of type Material Receipt, that returns that same item and Serial No to stock

The error message should be displayed on submit.

# Expected behaviour

The materials receipt should be allowed, i.e. the item's serial number should not be considered a duplicate.

# Suggested changes:

Modify "has_duplicate_serial_no()" to allow a materials receipt on a serial number that has previously been issued from stock through a Delivery Note (or Materials Issue).
